### PR TITLE
Fix Bug when Placing Transcription Lines on Touchscreen

### LINKED ISSUE #1640: Classifier: Transcription line tool is broken on touch screens
**Package**
lib-classifier

**Describe the bug**
Tapping twice on a touchscreen should place the start and end points of a transcription line, but only places the start point.

**Expected behavior**
The second `pointerDown` event on the drawing surface should place the end of the line, and mark it as finished (or open the subtask box, because the line isn't finished until text has been entered?)

**Additional context**
I've reproduced this on an iPhone, but it should happen on any touch device: iOS or Android. The underlying cause is that I wrote the `TranscriptionLine` tool so that it depends on `pointerMove` in order to place the end of the line. This works with a trackpad or mouse (`pointerDown -> pointerUp -> pointerMove -> pointerDown`), but not with touch (`pointerDown -> pointerUp -> pointerDown`.) The fix is probably to abstract event handling, during mark creation, so that `pointerDown` on the drawing surface is passed down to the active mark, which then figures out how to handle it.

### PR DESCRIPTION
**Pull request by wgranger**
_Please request review from `@zooniverse/frontend` team. If PR is related to design, please request review from `@beckyrother` in addition._ 

Package: lib-classifier

Closes #1640 

Describe your changes:
This PR isn't quite ready for a final review. This bug is very tricky to pen down, so I'm putting it up here for some feedback. Here's what I've attempted thus far:

1. Currently, a transcription line is finished with a pointer up [event](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/plugins/drawingTools/experimental/components/TranscriptionLine/TranscriptionLine.js#L116). However, this isn't triggered with a touchUp (for reasons mentioned [here](https://www.html5rocks.com/en/mobile/touchandmouse/#:~:text=The%20behavior%20of%20these%20events,currently%20under%20the%20mouse%20cursor)) as the SVG circle isn't visible until after a touch as occurred. This also sends a node/ref to tell the subTask where to position itself. 

1. Because of that, let's try to intercept the mark at the onPointerDown method [here](https://github.com/zooniverse/front-end-monorepo/blob/da4dce116fa696ca191ba8757ae3a5d0584835dc/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js#L108) to programmatically finish the transcription if we should (this needs cleaning up, needed to get a POC working first)

1. By doing this, we don't have access to the node, which tells us where to place the subTask. We can't access that node with touch events for reasons mentioned in the article linked above. Perhaps there's another way to determine placement of the box?

1. I tried creating a `setSubTaskVisibility` on the transcription line tool which sets the box based on `x1, x2, y1, y2` values of a mark. However, this doesn't quite work as expected because the subTask is usually set by checking the [bounding client rect](https://github.com/zooniverse/front-end-monorepo/blob/da4dce116fa696ca191ba8757ae3a5d0584835dc/packages/lib-classifier/src/plugins/drawingTools/models/marks/Mark/Mark.js#L94). I could perhaps get the correct values if I had access to the `subjectViewer.scale`, but I'm not 100% sure where that is kept in the store, or if that approach works.

Any suggestions here? I've seemingly been foiled at every turn in fixing this bug. However, it could be I'm overlooking something since I didn't implement the drawing tools. 

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by srallen** [comment]
I opened #1736 to add a suggested change to make sure the other tools stay working. I'm still unsure on approach here. The `TranscriptionLine` mark component _also_ has an `onPointerDown`  handler, which I think was added to fix another bug, and keeping track of all of these is getting hard to understand.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 1]
```diff
@@ -58,16 +58,23 @@ function InteractionLayer ({
       return true
     }
 
-    const activeMark = activeTool.createMark({
+    const activeMark = activeTool.onPointerDown({
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:61]: I think `createMark` should remain, but shouldn’t be called if `creating` is true. If we are already in the middle of creating a mark, then we delegate the pointerDown event to the active mark, which can then handle it if necessary.

If `creating` is false, then we call `activeTool.createMark()` and start a new mark.

**Comment by eatyourgreens** [comment]
If it’s useful, I added a Readme for the tool and mark APIs in `src/plugins/drawingTools`.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:61]: So maybe leave the original mark creation code in place, but put this before it. 

```js
if (creating) {
  return activeMark.handlePointerDown(event)
}
```

Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:61]: I like the idea of the active tool handling events during mark creation. That removes the ambiguity between tapping to place a transcription line and tapping on a transcription line to select and move it.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 2]
```diff
@@ -58,16 +58,23 @@ function InteractionLayer ({
...
     })
-    activeMark.initialPosition(convertEvent(event))
-    setActiveMark(activeMark)
-    setCreating(true)
-    activeMark.setSubTaskVisibility(false)
-    return false
+
+    if (activeMark.finished) {
+      onPointerMove(event)
+      onFinish(event)
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:69]: I think tying `onPointerDown` to finishing a mark will clash with the polygon tool, where you tap multiple times to create a shape. Behaviour for tools and marks should be in the code for those individual types of tools or marks so that it doesn’t interfere with other shapes. 

Having said that, I’ve tried AI4Mars on my phone and it is pretty much impossible to draw a shape with the current polygon tool.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js (hunk 1 above)
Comment (srallen) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:61]: I like the suggestions here. It's a bit more clear and keeps the mark creation decoupled from handling pointer down. I'd still suggest doing something like #1736 and have a generic handler in the base `Tool` model. I'll close it for now, but I'll leave the branch as a reference.

**Comment by wgranger** [comment]
This makes sense in handling the pointer down events, but a bigger issue of solving this bug is sending the correct svg line ref/node when the mark is finished, which currently happens [here](https://github.com/zooniverse/front-end-monorepo/blob/da4dce116fa696ca191ba8757ae3a5d0584835dc/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.js#L28). If the mark is finished programmatically with a `activeMark.handlePointerDown` event, it won't be possible to grab the SVG node to determine where to place the subtask.

**Comment by eatyourgreens** [comment]
I think the ref is added when the Mark component calls `props.onFinish`, and will point to the mark that invoked the callback.

**Comment by wgranger** [comment]
Yes, that's what currently happens on the [TranscriptionLine pointerUp](https://github.com/zooniverse/front-end-monorepo/blob/da4dce116fa696ca191ba8757ae3a5d0584835dc/packages/lib-classifier/src/plugins/drawingTools/experimental/components/TranscriptionLine/TranscriptionLine.js#L116), but that `pointerUp` event doesn't fire with a touch up, only a mouseUp/click. So, the idea is how to get the subtask placement without access to the SVG node.

**Comment by eatyourgreens** [comment]
The mark component could react to the mark model’s finished state changing from true to false. 

Fixing the sub task is out of the scope of my original issue, so maybe put that into its own issue and PR?

**Comment by srallen** [comment]
I'm supportive of keeping this fix atomic and addressing the subtask positioning separately.

**Comment by wgranger** [comment]
Cool. If that's the case, I think I can get a decent solution up with the subtask positioned at `X: 0, y: 0` and we can go from there.

**Comment by wgranger** [comment]
This utilizes a `handlePointerDown` even per suggestions above. Remaining issues include finishing a mark and showing the subtask.

A mark could be finished as follows (here is [context](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js#L81)):

```
if (creating) {
    let activeMark = activeTool.handlePointerDown && activeTool.handlePointerDown(convertEvent(event))
    return onFinish(activeMark)

}
```

Two issues remain here:
- Subtask will appear at `x: 0, y: 0` if it doesn't receive a node
- Subtask will close immediately after it opens because it assumes a click that is off the modal should close the modal

Happy to explore these issues here or leave their resolution for a separate PR.

**Comment by wgranger** [comment]
Alright! I've got a working solution here by checking if `mark.finished` changes and calling `onFinish` when it does if there isn't a visible subtask. I also called `setCreating(false)` in the interaction layer to trigger a rerender and allow the onFinish check to occur.

If this approach looks good I'll add some tests.

**Comment by srallen** [comment]
Gonna have another look this afternoon. Thanks for this.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 3]
```diff
@@ -57,12 +57,20 @@ function InteractionLayer ({
     if (disabled || move) {
       return true
     }
+    let activeMark
 
-    const activeMark = activeTool.createMark({
+    if (creating) {
+      activeMark = activeTool.handlePointerDown && activeTool.handlePointerDown(convertEvent(event))
+      if (activeMark.finished) setCreating(false)
+      return activeMark
```
[the remark this answers is not in the document]
Comment (wgranger) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:65]: There's no good reason for this. It should probably return a boolean

File: packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js [hunk 4]
```diff
@@ -66,6 +66,10 @@ const Tool = types.model('Tool', {
       self.marks.delete(mark.id)
     }
 
+    function handlePointerDown (mark) {
+      return self.createMark(mark)
```
[the remark this answers is not in the document]
Comment (wgranger) [reply packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:70]: I'm thinking this is still needed for something like a polygon tool using a touch screen. If using a polygon tool, a second mark placed will probably hit the 
`activeTool.handlePointerDown && activeTool.handlePointerDown(convertEvent(event))` line, which will fail if that function is not defined on the tool.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 5]
```diff
@@ -78,7 +86,7 @@ function InteractionLayer ({
     }
   }
 
-  function onFinish (event, node) {
+  function onFinish (event = {}, node) {
```
[the remark this answers is not in the document]
Comment (wgranger) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:89]: Yes, this will fail without a default parameter because of the line underneath it, `const { target, pointerId } = event`. This function will be called without an event on the `useEffect` change made in this PR.

**Comment by wgranger** [comment]
I did intend to add tests/specs to this PR, I just wanted to make sure this approach looked good before writing the specs. If this solution seems to work then my next steps are to add specs.

**Comment by srallen** [comment]
@wgranger Apologies on my part. I misinterpreted the most recent request for review as a request for final review rather than continued request for comment on approach. I think the approach here is a viable solution and it's readable in its approach. I had a few questions which you've answered, so I would say that this is ready for specs. 👍

File: packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:70]: That line checks if `activeTool.handlePointerDown` exists, so won’t fail if it isn’t defined.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:89]: `onFinish` is an event handler that's passed as a prop to individual mark components and then called when you finish drawing. https://github.com/zooniverse/front-end-monorepo/tree/master/packages/lib-classifier/src/plugins/drawingTools#react-components

Here's a couple of examples where it's passed a pointer event:

https://github.com/zooniverse/front-end-monorepo/blob/661bc16f39297d1eb5fc4bf5555515d973c0ac68/packages/lib-classifier/src/plugins/drawingTools/components/Point/Point.js#L25-L27

https://github.com/zooniverse/front-end-monorepo/blob/661bc16f39297d1eb5fc4bf5555515d973c0ac68/packages/lib-classifier/src/plugins/drawingTools/components/Ellipse/Ellipse.js#L25-L27

`useEffect` here is using `onFinish` in a slightly different manner than it was originally intended to be used, but I'm not sure if that's really a big deal, or if the original idea of it as an event handler is useful when we're dealing with multiple gestures to create a shape.

File: packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (wgranger) [reply packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:70]: But will it still add a mark for something like a polygon if using a touch screen? I believe using a touchscreen would call `onPointerDown` in the `InteractionLayer` with each touch, which would essentially do nothing for a polygon without this fallback.

**Comment by wgranger** [comment]
Alright. Got some tests in here, including some missing `TranscriptionLineTool` tests. I believe the `handlePointerDown` fallback is still necessary for the `Tool` model.

File: packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:70]: `createMark()` starts a new Mark, so it would place the first point for a new Polygon. TranscriptionLine is a special case of Polygon with only two points, I suppose.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.spec.js [hunk 6]
```diff
@@ -184,6 +186,22 @@ describe('Component > InteractionLayer', function () {
...
+            type: 'pointer',
+            target: {
+              setPointerCapture: sinon.stub(),
+              releasePointerCapture: sinon.stub()
+            }
+          }
+          wrapper.find(StyledRect).simulate('pointerdown', fakeEvent)
+          wrapper.find(StyledRect).simulate('pointerdown', fakeEvent)
+          expect(activeTool.handlePointerDown).to.have.been.calledOnce()
+        })
```
Comment (srallen) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.spec.js:203]: What this spec is testing is slightly confusing. I think this is testing that `handlePointerDown` is called only once, which is why `pointerdown` is simulated twice? Perhaps the description should be `should call the handlePointerDown function only once`?

File: packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (srallen) [reply packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:70]: Yeah, I think we need to pick a single code path here:

- Option 1: The `InteractionLayer` does an existential check for the existence of the `handlePointerDown` action before it calls it. If it exists on the current active tool and the mark is `creating`, then it is called and the `onPointerDown` function ends with returning true. If it does not exist and the mark is not `creating`, `createMark` is called by the same function in the `InteractionLayer` and then sets the new mark as active, its initial position, sets the subtask visibility and returns false.
- Option 2: The `InteractionLayer` does not do an existential check and we have the generic `handlePointerDown` action defined on the generic `Tool` model that all tools are composed with. The generic version of this handler calls createMark and then handles any other actions needing to be done like setting the active mark, etc.

As it is now, the code path can be a little hard to follow since there's two paths for how a mark might get created and it's not clear what the conditions are for when one happens vs the other (an active tool's `handlePointerDown` creates the mark vs the `onPointerDown` in `InteractionLayer`). Is there a chance `createMark` gets called twice when we don't want it to? I'm not sure. 

This further refactoring, though, I think is out of scope of this fix, but I'd like to see it at some point.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (srallen) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:89]: I would suggest that we have an `onPointerUp` event handler and a separate `onFinish` function and decouple what's going on here. That's out of scope for this fix though.

File: packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:70]: Just a note that returning true from an event handler historically signals that the browser should continue to handle the event with its default handler. So, I think we want to return false except in the default case, where the interaction layer ignores the event. 

I don't think any modern browser uses Boolean returns from a handler, but preventing the default action and also returning true can be confusing to read.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:89]: I think I might have chosen bad examples that both use `onPointerUp` to finish a mark. `onFinish` was added as a Mark prop so that we have the decoupling that @srallen mentions. Finishing an Ellipse on pointerUp was definitely a mistake, because you really want to wait until both axes have been drawn before finishing the shape.

Prior to that, the interaction layer had an `onPointerUp` handler, but then the drawing surface is dictating how every mark, regardless of type, should handle events. That breaks when you want a mark that doesn't conform to the generic mark behaviour eg. a line that's drawn by tapping twice on the drawing canvas.

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:89]: I think the piece that still needs to be decoupled is setting and releasing pointer capture for multiple gestures. I don’t think that needs to happen in this PR, but `onFinish()` in this code will not manually release a pointer, which means this PR relies on the browser to release any captured pointers. 

I think `setPointerCapture` gets called during the initial pointerDown. Browsers should release it (I think) on the next pointerUp emitted with that pointerID.

[the remark this answers is not in the document]
Comment (srallen) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:89]: I think generally speaking that this is all a bit too abstracted. What I'd like to see is that each tool has its own set of pointer down, pointer up, etc handlers and yes, I realize that this might mean some duplication of code, but we'll buy back some flexibility and clarity about what each tool does for each event. Then we can decide on how to abstract it to some general handlers if needed. I'd also like to see it consolidated a bit: either let's define these in the models or let's define them in the components. Right now we have pointer event handling spread across MST models and components and again, it's not clear what gets called when. Out of scope for this though.

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:89]: Sure. I was more concerned that this block isn't being called so wanted to be absolutely sure we aren't capturing the pointer then not releasing it.
https://github.com/zooniverse/front-end-monorepo/blob/f0f63edeafaabb7c7126197fc3def7194b4c2ada/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js#L96-L98

**Comment by wgranger** [comment]
Updated tests using a test wrapper so I won't have to rewrite the entire wrapper setup with each test. I can also add the most recent comments into a separate issue concerning refactoring so we have documentation of next steps for this feature.

File: packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:70]: > Is there a chance createMark gets called twice when we don't want it to? I'm not sure.

I think it does. My understanding is that this handler only gets called when `creating` is true, so we would have mark creation happening twice ie. creating two new ellipses or points or lines when we only want one. In actuality, `pointerUp` cancels creation for each of those shapes so this handler is never called. It will be called when we build tools that don't end creation on pointerUp eg. polygons or maybe freehand drawing. Ellipses probably shouldn't end creation on pointerUp either. I think that was a mistake on my part.

If there isn't a test already, I think there should be a test that pointerDown on the interaction layer does not create a new mark when `creating` is true.

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:70]: Following up on @srallen's comments about changing the code paths here, which is confusing, any change to the drawing API should also be documented in the README for the drawing tools.

**Review by eatyourgreens — CHANGES_REQUESTED** [review]
One small change: You've accidentally committed a .DS_Store file from OSX.

**Comment by wgranger** [comment]
@srallen Yep. I can create a separate issue for those notes
